model/scope.go

```diff
@@ -0,0 +1,168 @@
+package model
+
+import (
+	"encoding/json"
+	"strings"
+)
+
+type Scope []string
+
+const (
+	ScopeChannelJoin  = "channels:join"
+	ScopeMessageRead  = "message:read"
+	ScopeMessageWrite = "message:write"
+	ScopeFilesRead    = "files:read"
+	ScopeFilesWrite   = "files:write"
+	ScopeSearchPosts  = "search:post"
+)
+
+func ScopeAny(scopes ...string) Scope {
+	return scopes
+}
+
+func ScopeAllow() Scope {
+	return Scope{}
+}
+
+func ScopeDeny() Scope {
+	return nil
+}
+
+func (s Scope) IsInScope(scope string) bool {
+	if s == nil {
+		return false
+	}
+
+	for _, allowed := range s {
+		if scope == allowed {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s Scope) IsPluginInScope(pluginID string) bool {
+	if s == nil {
+		return false
+	}
+
+	for _, allowed := range s {
+		if strings.HasPrefix(allowed, "plugin:") &&
+			strings.TrimPrefix(allowed, "plugin:") == pluginID {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s Scope) Intersection(scope Scope) Scope {
+	if s == nil || scope == nil {
+		return ScopeDeny()
+	}
+
+	out := Scope{}
+	for _, x := range s {
+		if scope.IsInScope(x) {
+			out = append(out, x)
+		}
+	}
+
+	if len(out) == 0 {
+		return ScopeDeny()
+	}
+
+	return out
+}
+
+func (s Scope) ToJSON() string {
+	bytes, err := json.Marshal(s)
+	if err != nil {
+		return ""
+	}
+	return string(bytes)
+}
```

Comment:
I keep postponing the announcement, but we have got rid of all `ToJSON` methods in the codebase. Just call json.Marshal directly and _always_ handle the error.